lib/Sema/CSGen.cpp

```diff
@@ -3045,7 +3045,12 @@ namespace {
                                          /*decl*/ nullptr, locator);
           break;
         }
-        
+
+        case KeyPathExpr::Component::Kind::TupleElement: {
+          llvm_unreachable("[technicated]");
```

Comment:
Actually I'm having a hard time triggering this case to even execute. I tried to set up various error scenarios, but this unreachable is never run... I can see that the kind(s) of the KP component in this context are `UnresolvedProperty`, `UnresolvedSubscript`, `OptionalChain`, etc., but never `TupleElement`. I also never see `Property` or `Subscript`, so I think that the reason the code never executes could be that `TupleElement` is resolved starting from an `UnresolvedProperty` (in `CSApply::visitKeyPathExpr`). If this is true, how can one ever make this `case` execute?

Maybe simply the code I came up with is not enough to cause the behavior?

```
// compiled using:
// swift -frontend -c -primary-file /path/to/this/file.swift -o /wathever

struct S {
  var a = 99
}

class C {
  var a: Int = 101
  subscript(k: String) -> Int {
    return -1
  }
}

struct Test {
  var s = S()
  var a = (1, "Hello", (99, 98))
  var b = 3
  var c = (x: 4, y: 5, z: C())
}

func test1(_ kp: KeyPath<Test, Int>) {
  let t = Test()
  print(kp, t[keyPath: kp])
}

@discardableResult
func test2<T>(_ kp: KeyPath<Test, T>) -> T {
  let t = Test()
  return t[keyPath: kp]
}

func test3(_ kp: KeyPath<Test, Int?>) {
  let t = Test()
  print(kp, t[keyPath: kp] as Any)
}

test1(\Test.a.0)
//test1(\Test.a.1)
test1(\Test.c.x)
//test1(\Test.c.z)
test1(\Test.c.z.a)

test2(\Test.a.0)
test2(\Test.a.1)
test2(\Test.c.x)
test2(\Test.c.z)
test2(\Test.c.z.a)
test2(\Test.c.z["any_str"])

//test3(\Test.a.0)
//test3(\Test.a.1)
//test3(\Test.c.x)
//test3(\Test.c.z)
//test3(\Test.c.z.a?)

extension Test {
  var maybe: (Int, C)? {
    get { return nil }
    set { }
  }
}

//test3(\Test.maybe)
//test3(\Test.maybe.0)
test3(\Test.maybe?.1.a)
```

Reply:
@xedin might be able to help suggest a test case. AIUI, it would trigger when the constraint system for the types fails to find an answer, so maybe a deliberately wrong type annotation could trigger it, like:

```
let x: KeyPath<(Int, String), Int> = \.1 // ought to be KeyPath<(Int, String), String>
```

Reply:
It's generally hard to come up with test-cases for things like that but I think in this case you might try something like - a call with keypath expression as an argument where it refers to a tuple and give it incorrect contextual result type. I think that should do the trick because diagnostics would attempt to re-typecheck expression with and without contextual type which would mutate AST and introduce tuple element e.g.

```swift
func foo(_ keypath: KeyPath(Int, String), Int>) -> Int { return ... }

let _: String = foo(\.0)
```

Reply:
Unfortunately that did not work - I got this error instead:

```swift
<REPL Input>:1:17: error: cannot convert value of type 'Int' to specified type 'String'
let _: String = foo(\.0)
                ^~~~~~~~
```

But luckily your suggestion drove me to try this code:

```swift
func foo<T>(_: KeyPath<(T, String), T>) -> Int { ... whatever ... }
foo(\(Int, Int).0) // CRASH!
```

So thank you very much for you support!

Reply:
@jckarter I'm ashamed to ask a question like this, but after some days of trial and error I have come to the conclusion that I do not know how to proceed here. I think I understand what I should do but I don't know how to express it in code so... what should I do?

I thought I should make `base` bindable to the `Type` of the tuple element at the given index and then set `base` to be it, but I cannot even get to the `TupleType` because I have only access to what I understand to be `TypeVariableType`s. 😔